### ISSUE TITLE
CRITICAL: Personalization still producing zero/negative gains (Issue #43 not resolved)

### ISSUE THREAD

**Issue by reinesaj2** [reinesaj2/federated-ids#163]
## Problem Statement

Issue #43 was closed as resolved, but **analysis of 1,600 clients from 195 experiments shows personalization is STILL not working**.

## Evidence from Latest Analysis (2025-11-21)

```
Overall mean gain:        -0.000018  (essentially zero, slightly NEGATIVE)
Median gain:               0.0000    (no improvement)  
Max gain observed:         0.0051    (0.5% - negligible)
Min gain:                 -0.0135    (-1.35% degradation)
Clients with gains >1%:    0.0%      (ZERO clients benefited!)
```

### Gains by Personalization Epochs:
- **3 epochs**: mean gain = **-0.00071** (WORSE than global model!)
- **5 epochs**: mean gain = **-0.00015** (WORSE than global model!)
- **0 epochs** (baseline): mean gain = 0.0

### Dataset Breakdown:
- Analyzed: 1,600 clients across COMP dataset
- Alpha values: 0.02, 0.05, 0.1, 0.2, 0.5, 1.0, Inf
- **NONE showed meaningful positive gains**

### 1. Implementation Bug (Most Likely)
- Personalization logic may not be executing
- Model parameters not actually fine-tuning
- Clients testing same global model twice
- **Evidence**: Exactly 0.0 median across 1,600 samples is statistically impossible if personalization varied

### 2. Catastrophic Forgetting
- Fine-tuning destroying global knowledge
- Learning rate too high for local adaptation
- **Evidence**: Negative mean gains (-0.00018) suggest harm, not help

### 3. Test Set Contamination
- Personalized and global models evaluated on same data
- Gain calculation bug: `gain = personalized_f1 - global_f1` computing identical values
- **Evidence**: 0.0% positive gains means NO clients benefited at all

### 4. Insufficient Local Data
- Clients lack enough samples for meaningful adaptation
- Need to verify `len(local_test_set)` per client

### Immediate (P0):
- [ ] Add debug logging in `client.py` personalization phase:
  - Print model weight norms before/after fine-tuning
  - Log optimizer state and learning rate
  - Verify test set differs between global eval and personalized eval
- [ ] Run single-client experiment with verbose logging:
  ```bash
  python run_experiment.py --personalization_epochs=5 --alpha=0.1 --seed=42 --debug
  ```
- [ ] Compare model checkpoints: `global_model.pth` vs `personalized_model.pth` (should differ!)

### Validation (P1):
- [ ] Unit test: Verify fine-tuning changes model parameters
  ```python
  assert not torch.equal(global_weights, personalized_weights)
  ```
- [ ] Integration test: Personalized model should improve on client's local distribution
- [ ] Metric audit: Verify gain calculation uses correct model/data pairs

### Documentation (P2):
- [ ] Document expected vs actual personalization behavior
- [ ] Add assertions to catch zero-gain scenarios in CI
- [ ] Update Issue #36 (personalization CI) to include gain validation

## Impact

**CRITICAL BLOCKER for Thesis Objective 3:**
- Current results **invalidate** personalization claims
- Cannot publish thesis with zero-gain personalization
- Need reproducible evidence of positive adaptation gains

## Acceptance Criteria

- [ ] Root cause identified and documented
- [ ] Fix implemented with unit/integration tests
- [ ] New experiments show `mean(gains) > 0.01` for non-IID data (α ≤ 0.2)
- [ ] At least 30% of clients show gains > 0.01 in heterogeneous scenarios
- [ ] CI validates personalization works on every merge

## Related Issues

- Issue #43 (marked closed but not resolved)
- Issue #13 (personalization mechanisms - may need revision)
- Issue #20 (personalization workflow - implementation may be broken)
- Issue #36 (personalization CI - cannot enable until this is fixed)

## Priority Justification

**P0 - Thesis Blocker**: Without working personalization, Objective 3 fails. This invalidates a core thesis contribution and must be resolved before defense.
